lib/spack/spack/test/stage.py

```diff
@@ -728,6 +736,52 @@ def test_first_accessible_path(self, tmpdir):
         # Cleanup
         shutil.rmtree(str(name))
 
+    def test_create_stage_root(self, tmpdir, no_path_access):
+        """Test _create_stage_root permissions."""
+        spack_dir = tmpdir.join('path')
+
+        # Ensure an OS Error is raised on the created, non-user directory
+        with pytest.raises(OSError) as exc_info:
+            spack.stage._create_stage_root(str(spack_dir))
+
+        assert exc_info.value.args[0] == errno.EACCES
+
+        # Ensure an OS Error is raised on a created, user directory
+        spack_dir.ensure(dir=True)
+        user_dir = spack_dir.join(getpass.getuser())
+        with pytest.raises(OSError) as exc_info:
+            spack.stage._create_stage_root(str(user_dir))
+
+        assert exc_info.value.args[0] == errno.EACCES
+
+    def test_create_stage_root_non_uid(self, tmpdir, monkeypatch):
+        """Test _create_stage_root with non-uid user dir."""
+        orig_stat = os.stat
+
+        class MinStat:
+            st_mode = -1
+            st_uid = -1
+
+        def _stat(path):
+            p_stat = orig_stat(path)
+
+            fake_stat = MinStat()
+            fake_stat.st_mode = p_stat.st_mode
+            return fake_stat
+
+        user_dir = tmpdir.join(getpass.getuser())
+        user_dir.ensure(dir=True)
+        user_path = str(user_dir)
+
+        monkeypatch.setattr(os, 'stat', _stat)
+        spack.stage._create_stage_root(user_path)
+
+        assert os.stat(user_path).st_uid != os.getuid()
+
+        # Restore the original os.stat function to ensure Python 3 works
+        # during test stage directory cleanup.
+        monkeypatch.setattr(os, 'stat', orig_stat)
```

Comment:
It does except, as alluded to in the comment, Python 3.5 and 3.6 fail during cleanup apparently due to a "temporary" API incompatilibity (see https://travis-ci.org/spack/spack/builds/586338772).

Reply:
I say "temporary" because, as shown in the build results, Python 3.7 does not incur the same failure.

Reply:
Based on looking at the linked error, I think we may want to avoid either

* running the mock_stage fixture
* monkeypatching `os.stat`

```
    @pytest.fixture(scope='function', autouse=True)
    def mock_stage(clear_stage_root, tmpdir_factory, request):
        """Establish the temporary build_stage for the mock archive."""
        # Workaround to skip mock_stage for 'nomockstage' test cases
        if 'nomockstage' not in request.keywords:
            new_stage = tmpdir_factory.mktemp('mock-stage')
            new_stage_path = str(new_stage)
            # Set test_stage_path as the default directory to use for test stages.
            current = spack.config.get('config:build_stage')
            spack.config.set('config',
                             {'build_stage': new_stage_path}, scope='user')
            # Ensure the source directory exists
            source_path = new_stage.join(spack.stage._source_path_subdir)
            source_path.ensure(dir=True)
            yield new_stage_path
            spack.config.set('config', {'build_stage': current}, scope='user')
            # Clean up the test stage directory
            if os.path.isdir(new_stage_path):
>               shutil.rmtree(new_stage_path)
```

It seems that the monkeypatch fixture runs before the mock_stage fixture in this context so at the time that `rmtree` is called (which delves into Python lib code) the resulting os.stat fails. I'm not sure what controls the ordering of the fixtures but it may be arbitrary and I'm not sure if we should depend on a consistent ordering.

IMO it would be best to figure out how to test this without monkeypatching `os.stat`: e.g. the `Stage` could accept an object that handles the `.stat` calls and this test could create a mock version of that object.

Reply:
At this point I'm skipping the `mock_stage` fixture to avoid the Python 3.5/3.6 problem.

Reply:
> What is the concern with monkeypatching os.stat in this way?

Mocking the output of library functions outside Spack can have unexpected consequences; this case was a less-insidious instance of that in the sense that the mocked function signature mismatched the true function signature. The system and surrounding libraries could alter their own behavior based on changing the behavior of this function in ways that are harder to track (e.g. fail but don't throw exceptions).

Overall I don't think we mock any other functions outside of Spack and now that I think on it more I feel that is a good practice so I am inclined to think this should work around the mocking of `os.stat`.

Reply:
Actually it appears `cmd/common/arguments.py` mocks `multiprocessing.cpu_count` and mocking  it is a documented use of monkeypatch (e.g., https://doc.pytest.org/en/2.8.7/monkeypatch.html). 

That said, I understand your point.

Reply:
Leveraged `nomockstage` to eliminate resetting `os.stat` in `test_create_stage_root_bad_uid`.  Also included a comment that `pytest 3.6.0 on` added a context function that can be used to narrow the scope of patching third-party functions.  The corresponding code was included as comments.

IMHO it would be ideal to be able to leverage the newer context `monkeypatch` function to help improve code coverage.

Reply:
Should this be a separate PR or integrated into #13065 ?